File: frontend/src/types/patterns.ts
// ── Pattern Analysis Types ──

export interface IssueItem {
  issue: string;
  count: number;
  pct: number;
}

export interface TypePatternDetail {
  korean_type: string;
  total_count: number;
  consultation_results: Record<string, number>;
  result_pct: Record<string, number>;
  common_issues: IssueItem[];
  avg_review_months: number | null;
  supplement_required_pct: number | null;
}

export interface PatternResponse {
  project_type: string;
  total_analyzed: number;
  overall_total: number;
  analysis_year_range: string;
  type_data: TypePatternDetail[];
}

export interface PredictedIssue {
  issue: string;
  probability_pct: number;
  past_count: number;
  total_in_type: number;
  description: string;
}

export interface RiskAssessment {
  location_type: string;
  risk_level: string;
  issue_probability: number;
  top_issues: string[];
  supplement_probability_pct: number;
}

export interface RemediationSuggestion {
  issue: string;
  common_remediation: string[];
  frequency: number;
}

export interface PredictionResponse {
  project_type: string;
  korean_type: string;
  total_in_type: number;
  predicted_issues: PredictedIssue[];
  consultation_prediction: Record<string, number>;
  avg_review_months: number | null;
  supplement_required_pct: number | null;
  risk_assessment: RiskAssessment | null;
  remediation_suggestions: RemediationSuggestion[];
}

export interface SuggestedRule {
  suggestion_id: string;
  title: string;
  source_type: string;
  issue: string;
  evidence: string;
  recommended_severity: string;
  rationale: string;
}

export interface SuggestedRulesResponse {
  suggestions: SuggestedRule[];
  total: number;
}

export interface OverallPatternSummary {
  total_count: number;
  year_range: string;
  analyzed_at: string;
  unique_biz_types: number;
  biz_type_distribution: Record<string, number>;
  step_distribution: Record<string, number>;
  available_types: string[];
}

File: backend/app/schemas/patterns.py
"""패턴 분석 API 응답 스키마."""

from pydantic import BaseModel, Field


class IssueItem(BaseModel):
    """개별 지적항목."""
    issue: str
    count: int = 0
    pct: float = 0.0


class TypePatternDetail(BaseModel):
    """한국어 사업유형별 상세 패턴."""
    korean_type: str
    total_count: int
    consultation_results: dict[str, int] = {}
    result_pct: dict[str, float] = {}
    common_issues: list[IssueItem] = []
    avg_review_months: float | None = None
    supplement_required_pct: float | None = None


class PatternResponse(BaseModel):
    """GET /api/patterns/{project_type} 응답."""
    project_type: str
    total_analyzed: int
    overall_total: int
    analysis_year_range: str = ""
    type_data: list[TypePatternDetail] = []


class PredictedIssue(BaseModel):
    """예상 지적항목."""
    issue: str
    probability_pct: float
    past_count: int
    total_in_type: int
    description: str


class RiskAssessment(BaseModel):
    """입지 리스크 평가."""
    location_type: str
    risk_level: str
    issue_probability: float
    top_issues: list[str] = []
    supplement_probability_pct: float


class RemediationSuggestion(BaseModel):
    """보완 패턴 제안."""
    issue: str
    common_remediation: list[str] = []
    frequency: float


class PredictionResponse(BaseModel):
    """GET /api/patterns/{project_type}/predict 응답."""
    project_type: str
    korean_type: str
    total_in_type: int
    predicted_issues: list[PredictedIssue] = []
    consultation_prediction: dict[str, float] = {}
    avg_review_months: float | None = None
    supplement_required_pct: float | None = None
    risk_assessment: RiskAssessment | None = None
    remediation_suggestions: list[RemediationSuggestion] = []


class SuggestedRule(BaseModel):
    """규칙 보강 제안."""
    suggestion_id: str
    title: str
    source_type: str
    issue: str
    evidence: str
    recommended_severity: str
    rationale: str


class SuggestedRulesResponse(BaseModel):
    """규칙 보강 제안 목록."""
    suggestions: list[SuggestedRule] = []
    total: int = 0


class OverallSummary(BaseModel):
    """전체 분석 데이터 요약."""
    total_count: int
    year_range: str = ""
    analyzed_at: str = ""
    unique_biz_types: int = 0
    biz_type_distribution: dict[str, int] = {}
    step_distribution: dict[str, int] = {}
    available_types: list[str] = []
